cdap-proto/src/main/java/co/cask/cdap/proto/security/Principal.java

```diff
@@ -38,8 +38,10 @@
 
   private final String name;
   private final PrincipalType type;
-  private final int hashCode;
   private final String kerberosPrincipal;
+  // This needs to be transient because we don't want this to be populated during deserialization since that
+  // value will not be provided by the client.
+  private transient Integer hashCode;
```

Comment:
If it is a primitive type, it is safe. But because it is an Integer, it can be assigned to an object before that object is fully initialized. At least that is how I remember it...